### ISSUE TITLE
How to make MPVMediaControl.exe process to close together with mpv?

### ISSUE THREAD

**Issue by KonoVitoDa** [datasone/MPVMediaControl#15]
After closing mpv, the process continues to run in the taskbar indefinitely, and this behavior is especially annoying because it makes previously opened files unable to be moved/deleted as Windows claims they are currently running.

**Comment by xathian** [comment]
Yeah this is the reason I stick with the now abandoned MPV-SMTC. It opens and closes with MPV, it's also a silent subprocess so it never needs to be in the task bar to begin with.

**Comment by KonoVitoDa** [comment]
> Yeah this is the reason I stick with the now abandoned MPV-SMTC. It opens and closes with MPV, it's also a silent subprocess so it never needs to be in the task bar to begin with.

It doesn't seem to work on Windows 11 :(

**Comment by conradsrc** [comment]
Also interested in this for the same reason. Constantly getting "cannot delete folder in use" errors from Windows while deleting and re-organizing videos because this was still open in the System Tray even though all instances of mpv were closed.

Looks like the [`ExitIfNoControllers()`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/notify_media.lua#L261) in `notify_media.lua` gets registered to run when mpv is closed. It does an IPC send that's received in [`PipeServer.cs`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/PipeServer.cs#L193). Considering the other IPC calls seem to be working correctly, I _assume_ it's getting to this case successfully.

Not at all familiar with C# or relevant Windows APIs but the [`Exit()`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/Program.cs#L101) function that's supposed to be called eventually is also used [here in `Program.cs`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/Program.cs#L52C42-L52C46) for the tray icon's context menu, which works as you'd expect. Therefore, I also _assume_ the `Exit()` function itself is also working correctly, if called.

The above mentioned [`setQuit`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/PipeServer.cs#L193) case:

1) Checks if the `"quit"` parameter is `true` (should be, [based on the IPC call](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/notify_media.lua#L265) in `notify_media.lua`).
2) Checks that `Program.AppContext`, `controller`, and `controller.File.Path` aren't null (seems like that should be true but idk?).
3) Sets the controller to stop if it isn't already (probably fine?).
4) Calls [`RemoveController(pid)`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/Program.cs#L73) (no idea if it's successfully doing so).
5) Calls [`ExitIfNoControllers()`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/Program.cs#L113) which should then do some async stuff to call [`Exit()`](https://github.com/datasone/MPVMediaControl/blob/af9695f98e8b221a31fa5ff4ec1b5893a524069c/MPVMediaControl/Program.cs#L101), but only if `_controllers.Count == 0`, which should be true **if** the previous step worked.

So maybe:

* Registering with mpv to run a lua function on shutdown isn't working?
* IPC send/receive failure?
* One of those variables is null?
* `RemoveController(pid)` isn't successfully doing so... maybe the `pid` isn't matching the one used when the `MediaController` was created?
* Something with the `async` or `await` in `ExitIfNoControllers()` that calls `Exit()`? Seems correct to my untrained eye.

Wish I was set up and qualified to help, proooobably not a difficult problem to debug (knock on wood).

**Comment by Toastbroti** [comment]
A quick fix for this without support for multiple mpv instances:

```lua
function on_quit()
    if shot_path then
        os.remove(shot_path)
    end
    write_to_socket("^[setQuit](pid=" .. pid .. ")(quit=true)(socket_name=" .. mpv_socket_name .. ")$")
    
    -- Force kill MPVMediaControl.exe
    mp.command_native({
        name = "subprocess",
        playback_only = false,
        args = {"taskkill", "/F", "/IM", "MPVMediaControl.exe"},
    })
end
```
